pulsar-broker/src/main/java/org/apache/pulsar/compaction/Compactor.java

```diff
@@ -0,0 +1,205 @@
+
+import org.apache.pulsar.common.api.Commands;
+import org.apache.pulsar.common.api.proto.PulsarApi.MessageMetadata;
+import org.apache.pulsar.client.api.MessageId;
+import org.apache.pulsar.client.api.PulsarClient;
+import org.apache.pulsar.client.api.RawReader;
+import org.apache.pulsar.client.api.RawMessage;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Compactor for Pulsar topics
+ *
+ * Compaction will go through the topic in two passes. The first pass
+ * selects latest offset for each key in the topic. Then the second pass
+ * writes these values to a ledger.
+ *
+ * <p>The two passes are required to avoid holding the payloads of each of
+ * the latest values in memory, as the payload can be many orders of
+ * magnitude larger than a message id.
+*/
+public class Compactor {
+    private static final Logger log = LoggerFactory.getLogger(Compactor.class);
+    private static final String COMPACTION_SUBSCRIPTION = "__compaction";
+    private static final int MAX_OUTSTANDING_READS = 500;
+    private static final int COMPACTED_TOPIC_ENSEMBLE_SIZE = 3;
+    private static final int COMPACTED_TOPIC_ACK_QUORUM_SIZE = 3;
+    private static final String COMPACTED_TOPIC_LEDGER_PROPERTY = "CompactedTopicLedger";
+    static byte[] COMPACTED_TOPIC_LEDGER_PASSWORD = COMPACTION_SUBSCRIPTION.getBytes();
+
+    private final ExecutorService executor;
+    private final PulsarClient pulsar;
+    private final BookKeeper bk;
+
+    public Compactor(PulsarClient pulsar,
+                     BookKeeper bk,
+                     ExecutorService executor) {
+        this.executor = executor;
+        this.pulsar = pulsar;
+        this.bk = bk;
+    }
+
+    public CompletableFuture<Long> compact(String topic) {
+        final CompletableFuture<Long> future
+            = new CompletableFuture<>();
+        executor.submit(new CompactionTask(topic, future));
+        return future;
+    }
+
+    private class CompactionTask implements Runnable {
+        final String topic;
+        final CompletableFuture<Long> future;
+
+        CompactionTask(String topic, CompletableFuture<Long> future) {
+            this.topic = topic;
+            this.future = future;
+        }
+
+        @Override
+        public void run() {
+            RawReader reader = null;
+            try {
+                reader = RawReader.create(pulsar, topic,
+                                          COMPACTION_SUBSCRIPTION).get(10, TimeUnit.SECONDS);
+                Map<String,MessageId> latestForKey = new HashMap<>();
+
+                RawMessage firstMessage = reader.readNextAsync().get(10, TimeUnit.SECONDS);
+                MessageId firstMessageId = firstMessage.getMessageId();
+                latestForKey.put(extractKey(firstMessage), firstMessageId);
+                MessageId latestMessageId = firstMessageId;
+
+                BlockingQueue<Future<RawMessage>> futures = new ArrayBlockingQueue<Future<RawMessage>>(
```

Comment:
I don't think you need to use a queue here. because the reader (the consumer underneath) does prefetch, you don't need to duplicate the logic.

Reply:
the reader is implemented using consumer, so there is already flow control applied there. so you don't need to redo the flow control here.

Reply:
The queue performs two functions. It allows us to keep track of futures (so they can be cancelled) and it allows us to limit outstanding requests. Consumer uses a ConcurrentLinkedQueue internally, so you could issue reads forever (or until memory filled up).

Reply:
> keep track of futures (so they can be cancelled)

you need to cancel them because the way how you implement it - sending multiple readNext() calls. Instead, if you rely on pulsar itself (to do the prefetching and rate limiting), you don't need to track futures.

your logic can be much simpler as below:

for the first pass,

```
while (true) {
     if (compactFuture.isDone()) {  // `compactFuture` is the future returned by compact call, to allow the task to be cancelled.
          // break since the task is cancelled.
          break;
     }
     Message msg = reader.readNext().get(...);
}
```

for the second pass,

```
Semaphore sem = new Semaphore(OUTSTANDING_WRITES);
while (true) {
    sem.acquire();
    Message msg = reader.readNext().get(...);

    lh.asynAddEntry(..., (...) -> {
        sem.release(); // release semaphore after an entry is written.
    });
}
```

you can do this in async way:

```
CompletableFuture<Message> readOne() {
    
    return reader.readNextAsync()
       .thenCompose(message -> {

            // process the message.

             if (compactFututre.isDone()) {
                // break since the task is cancelled.
                return FutureUtil.exeception(...);
            } else {
                // continue
               return readOne();
            }
       });

}
```

Reply:
```
CompletableFuture<Message> readOne() {
    
    return reader.readNextAsync()
       .thenCompose(message -> {

            // process the message.

             if (compactFututre.isDone()) {
                // break since the task is cancelled.
                return FutureUtil.exeception(...);
            } else {
                // continue
               return readOne();
            }
       });

}
```
This is the pattern I'm thinking will be bad for GC. 

readOne() returns a future (future A). This future has a callback which when it runs will call readOne() creating another future, future B. future A will hold a reference to future B. future B's callback calls readOne(), creating another future, future C. B holds a reference to C. And so on.

Reply:
yes. thenCompose chains the dependents. it doesn't work as 'flatMap' at scala/twitter future. 

in this case, the result what we need is the `compactFuture`, you can just use `whenComplete` or `handleAsync` on handling read callback

```
    void readOne(Reader reader, CompletableFuture<Integer> compactFuture) {
       reader.readNext().whenCompleteAsync((v, cause) -> {
            if (null != cause) {
                doneFuture.completeExceptionally(cause);
            } else {
                if (compactFuture.isDone()) {
                   // if the compact future is cancelled ...
                } else {
                   readOne(reader, compactFuture);
                }
            }
        });
    }

```